File: src/nylon/parser/NylonParser.java
package nylon.parser;

import nylon.Builtins;
import nylon.InlineFunction;
import nylon.nylonobjects.NylonFunction;
import nylon.parser.parsers.*;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Supplier;

/**
 * Created by Aedan Smith.
 */

public class NylonParser {
    public ArrayList<BiFunction<StringIterator, NylonParser, NylonFunction>> parsers = new ArrayList<>(256);

    public HashMap<String, Supplier<NylonFunction>> functions = new HashMap<>();

    public NylonParser(File stdl) {
        this(
                stdl,
                DefaultParser::addTo,
                CaptureParser::addTo,
                CastParser::addTo,
                CharacterParser::addTo,
                ExternalFunctionParser::addTo,
                ForLoopParser::addTo,
                IfStatementParser::addTo,
                MacroParser::addTo,
                NumberParser::addTo,
                NylonFunctionParser::addTo,
                StringParser::addTo,
                WhileNotEmptyLoop::addTo
        );
    }

    @SafeVarargs
    public NylonParser(File stdl, Consumer<ArrayList<BiFunction<StringIterator, NylonParser, NylonFunction>>>... parsers) {
        for (Consumer<ArrayList<BiFunction<StringIterator, NylonParser, NylonFunction>>> parser : parsers) {
            parser.accept(this.parsers);
        }
        Builtins.build(this, stdl);
    }

    public void parse(StringIterator in, InlineFunction inlineFunction) {
        while (in.hasNext()) {
            NylonFunction parse = parse(in);
            if (parse != null) {
                inlineFunction.functions.add(parse);
            }
        }
    }

    public NylonFunction parse(StringIterator in) {
        NylonFunction nylonFunction = null;
        while (in.hasNext() && nylonFunction == null) {
            if (in.hasNext(2) && Objects.equals(in.peekString(2), "//"))
                in.until('\n');
            nylonFunction = parsers.get(in.peek()).apply(in, this);
        }
        return nylonFunction;
    }

    public void parseUntil(InlineFunction inlineFunction, StringIterator in, char c) {
        loop:
        while (in.hasNext()) {
            NylonFunction nylonFunction = null;
            while (in.hasNext() && nylonFunction == null) {
                if (in.hasNext(2) && Objects.equals(in.peekString(2), "//"))
                    in.until('\n');
                if (in.peek() == c)
                    break loop;
                nylonFunction = parsers.get(in.peek()).apply(in, this);
            }
            if (nylonFunction != null) {
                inlineFunction.functions.add(nylonFunction);
            }
        }
    }
}

File: src/nylon/parser/NylonParser.kt
package nylon.parser

import nylon.NylonFunction
import nylon.buildBuiltins
import nylon.concatenate
import nylon.parser.parsers.*
import java.io.Reader
import java.util.*
import java.util.function.BiFunction
import java.util.function.Consumer
import java.util.function.Supplier

/**
 * Created by Aedan Smith.
 */

interface Parser : BiFunction<CharIterator, NylonParser, NylonFunction?>

interface ParserBuilder : Consumer<HashMap<Char, Parser>>

class NylonParser(vararg builders: ParserBuilder) {
    val parsers = HashMap<Char, Parser>(256)

    val functions = HashMap<String, Supplier<NylonFunction>>()

    constructor() : this(
            DefaultParserBuilder(),
            NumberParserBuilder(),
            StringParserBuilder(),
            CharacterParserBuilder(),
            FunctionParserBuilder(),
            IfStatementParserBuilder(),
            FunctionCallParserBuilder(),
            ArgumentParserBuilder(),
            MacroParserBuilder()
    ) {
        buildBuiltins(this)
    }

    init {
        builders.forEach { it.accept(parsers) }
    }

    fun parseAll(src: String) = parseAll(CharIterator(src))

    fun parseAll(src: CharIterator): NylonFunction {
        val functions = ArrayList<NylonFunction>()
        while (true) {
            val nylonFunction = parse(src)
            if (nylonFunction != null) {
                functions.add(nylonFunction)
            } else {
                break
            }
        }
        return concatenate(functions)
    }

    fun parse(src: CharIterator): NylonFunction? {
        while (true) {
            src.skipUntilStatement()
            if (!src.hasNext())
                break
            val nylonFunction = parsers[src.peek()]!!.apply(src, this)
            if (nylonFunction != null) {
                return nylonFunction
            }
        }
        return null
    }

    fun parseUntil(src: CharIterator, c: Char): NylonFunction? {
        val nylonFunctions = ArrayList<NylonFunction>()
        while (src.hasNext()) {
            if (!src.hasNext() || src.peek() == c)
                break
            val nylonFunction = parsers[src.peek()]!!.apply(src, this)
            if (nylonFunction != null) {
                nylonFunctions.add(nylonFunction)
            }
        }
        if (src.hasNext())
            src.next()
        return concatenate(nylonFunctions)
    }
}

fun CharIterator.skipUntilStatement() {
    while (hasNext() && (peek() <= ' ' || peek() == '/')) {
        if (peek() == '/') {
            if (hasNext(1) && peek(1) == '/') {
                while (hasNext() && next() != '\n') {
                }
                continue
            } else return
        }
        next()
    }
}

fun Reader.readExtendedText(): String {
    var s = ""
    while (ready()) {
        val c = read().toChar()
        if (c == '\\') {
            val n = read().toChar().toString() + read().toChar().toString()
            s += when (n) {
                "al" -> 'À'
                "ar" -> 'Á'
                "cc" -> 'â'
                "cd" -> 'à'
                "cf" -> 'å'
                "cl" -> 'ä'
                "cn" -> 'á'
                "cs" -> 'ã'
                "fn" -> 'ƒ'
                "ld" -> 'ï'
                "lp" -> 'ì'
                "lh" -> 'î'
                "lc" -> 'í'
                "ps" -> 'ð'
                "mt" -> 'ù'
                "mb" -> 'ú'
                "ss" -> 'Ï'
                "uq" -> '¿'
                else -> "\\" + n
            }
        } else {
            s += c
        }
    }
    return s
}
